ls1intum/Artemis
# `Discussion`: Reference artifacts within postings

### Checklist
#### General



#### Client

### Motivation and Context
In the current context, users refer to artifacts of Artemis such as lectures, attachments, and exercises byword in their postings. Due to this, others who want to get an idea of the mentioned artifact within a discussion first have to manually find and navigate to the artifact to access it. 

### Description
To enhance the user experience, we introduce two dropdowns to the markup editor of postings, 'Exercises' and 'Lectures'. When clicked, the list of the Exercises and Lectures of the current course is displayed. Attachments are also displayed in a nested way, where each is located under their specific Lecture. When a user selects an artifact, the corresponding markdown code is automatically generated and inserted into the markdown editor. Later when in preview mode or the posting is just viewed, the attachment's name is displayed as a link, together with an icon with respect to the type of the attachment.

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Course with postings enabled

1. Log in to Artemis
2. Navigate to Course Discussions page
3. Create/edit postings with references to types (check the screencast for help)
- Post (can reference with : #{POST_ID} )
- File-upload Exercise
- Programming Exercise
- Modeling Exercise
- Text Exercise
- Quiz Exercise
- Lecture
- Lecture Attachment
4.  Verify that unique icons are displayed for each type of artifact referenced and make sure that they are consistent with the icons used in Artemis for those types.
5. Verify that previewing within the markdown editor works as well.
6. Make sure that all types of references are accessed when clicked.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
  exerciseReferenceCommand.ts | 100% | 100% |                                                                
  lectureAttachmentReferenceCommand.ts | 90% | 100% |
  metis.service.ts | 78.7% |  92.54% |
  metis.util.ts | 100% | 100% |                                                          
  posting-content.components.ts | 82.45% | 97.95% |                                    
  posting-content-part.components.ts | 91.66% | 100% | 
  posting-markdown-editor.component.ts | 77.77% | 92.85% |                                    
  negated-type-check.pipe.ts | 100% | 100% |                                                          





### Screenshots (UPDATED!!!)


https://user-images.githubusercontent.com/67344158/174290816-0e3c5777-36ab-417e-aac6-b94ff07af3bb.mp4
